src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -156,39 +170,133 @@ public UnfilteredRowIterator queryStorage(PrimaryKey key, ReadExecutionControlle
      * <p>
      * Each (expression, SSTable indexes) pair is then passed to
-     * {@link IndexSearchResultIterator#build(Expression, Collection, AbstractBounds, QueryContext)}
+     * {@link CheckpointingIterator#build(Expression, Collection, AbstractBounds, QueryContext)}
      * to search the in-memory index associated with the expression and the SSTable indexes, the results of
      * which are unioned and returned.
      * <p>
-     * The results from each call to {@link IndexSearchResultIterator#build(Expression, Collection, AbstractBounds, QueryContext)}
+     * The results from each call to {@link CheckpointingIterator#build(Expression, Collection, AbstractBounds, QueryContext)}
      * are added to a {@link KeyRangeIntersectionIterator} and returned.
      */
-    public KeyRangeIterator.Builder getIndexQueryResults(Collection<Expression> expressions)
+    public KeyRangeIterator getIndexQueryResults(Collection<Expression> expressions)
     {
-        KeyRangeIterator.Builder builder = KeyRangeIntersectionIterator.builder(expressions.size());
+        // FIXME the ANN expression should move to ORDER BY, maybe something like:
+        // SELECT * FROM foo ORDER BY columnname ANN OF ?
+        var annExpression = getAnnExpression(expressions);
+        if (annExpression != null)
+            expressions = expressions.stream().filter(e -> e != annExpression).collect(Collectors.toList());
 
-        QueryViewBuilder queryViewBuilder = new QueryViewBuilder(expressions, mergeRange);
-
-        Collection<Pair<Expression, Collection<SSTableIndex>>> queryView = queryViewBuilder.build();
+        var queryView = new QueryViewBuilder(expressions, mergeRange).build();
+        Map<Memtable, List<KeyRangeIterator>> iteratorsByMemtable = expressions
+                                                                    .stream()
+                                                                    .flatMap(expr -> {
+            var mim = expr.context.getMemtableIndexManager();
+            return mim.iteratorsForSearch(expr, mergeRange, getLimit()).stream();
+        }).collect(Collectors.groupingBy(pair -> pair.left,
+                                         Collectors.mapping(pair -> pair.right, Collectors.toList())));
 
         try
         {
-            for (Pair<Expression, Collection<SSTableIndex>> queryViewPair : queryView)
-            {
-                @SuppressWarnings({"resource", "RedundantSuppression"}) // RangeIterators are closed by releaseIndexes
-                KeyRangeIterator index = IndexSearchResultIterator.build(queryViewPair.left, queryViewPair.right, mergeRange, queryContext, command.limits().count());
+            var sstableIntersections = queryView.view.entrySet()
+                                       .stream()
+                                       .map(e -> {
+                                           var it = createIntersectionIterator(e.getValue());
```

Comment:
Some ideas for speeding up requests when non-ann index is too restrictive or matches most data:
-  `IntersectionStatistics` needs to be fixed to use `count = Math.min(count, range.getCount();)`
- `sstableIntersections#getCount()` should provide a decent estimate about the potential matching for the sstable
  -  restrictive: when `count / sstable-rows` is small, e.g. 10% or count is less than limit, it will skip ANN index and leave top-k to top-k processor
  - match most data: when `count / sstable-rows` is large, e.g. 80%, or count is much larger than limit, it will skip iterating `sstableIntersections` and execute ANN search with a new limit which is `limit * sstable-rows / count`.

Reply:
Made a note for this too